sensors/odometer.go

```diff
@@ -0,0 +1,183 @@
+// Package sensors defines interfaces for sensors used by viam cartographer
+package sensors
+
+import (
+	"context"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/edaniels/golog"
+	geo "github.com/kellydunn/golang-geo"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+	"go.viam.com/rdk/components/movementsensor"
+	"go.viam.com/rdk/components/movementsensor/replay"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/utils/contextutils"
+	goutils "go.viam.com/utils"
+)
+
+// TimedOdometerSensor describes a sensor that reports the time the reading is from & whether or not it is
+// from a replay sensor.
+type TimedOdometerSensor interface {
+	Name() string
+	DataFrequencyHz() int
+	TimedOdometerSensorReading(ctx context.Context) (TimedOdometerSensorReadingResponse, error)
+}
+
+// TimedOdometerSensorReadingResponse represents an odometer sensor reading with a time & allows the caller
+// to know if the reading is from a replay movement sensor.
+type TimedOdometerSensorReadingResponse struct {
+	Position    *geo.Point
+	Orientation spatialmath.Orientation
+	ReadingTime time.Time
+	Replay      bool
+}
+
+// Odometer represents an Odometer movement sensor.
+type Odometer struct {
+	name            string
+	dataFrequencyHz int
+	Odometer        movementsensor.MovementSensor
+}
+
+// Name returns the name of the Odometer.
+func (odom Odometer) Name() string {
+	return odom.name
+}
+
+// DataFrequencyHz returns the data rate in ms of the Odometer.
+func (odom Odometer) DataFrequencyHz() int {
+	return odom.dataFrequencyHz
+}
+
+// TimedOdometerSensorReading returns data from the odometer movement sensor and the time the reading is from.
+func (odom Odometer) TimedOdometerSensorReading(ctx context.Context) (TimedOdometerSensorReadingResponse, error) {
+	replay := false
+
+	var readingTimePosition, readingTimeOrientation time.Time
+	var position *geo.Point
+	var orientation spatialmath.Orientation
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return TimedOdometerSensorReadingResponse{}, nil
+		default:
+			if readingTimePosition == defaultTime || readingTimePosition.Sub(readingTimeOrientation).Milliseconds() < 0 {
+				ctxWithMetadata, md := contextutils.ContextWithMetadata(ctx)
+				if position, _, err = odom.Odometer.Position(ctxWithMetadata, make(map[string]interface{})); err != nil {
+					return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, "Position error")
+				}
+
+				readingTimePosition = time.Now().UTC()
+				if timeRequestedMetadata, ok := md[contextutils.TimeRequestedMetadataKey]; ok {
+					replay = true
+					if readingTimePosition, err = time.Parse(time.RFC3339Nano, timeRequestedMetadata[0]); err != nil {
+						return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, replayTimestampErrorMessage)
+					}
+				}
+			}
+
+			if readingTimeOrientation == defaultTime || readingTimeOrientation.Sub(readingTimePosition).Milliseconds() < 0 {
```

Comment:
Should we check if the times are within `replayTimeToleranceMsec` before making this call? Otherwise we might keep updating both position and orientation indefinitely.